book/core_programming.md

```diff
@@ -2,153 +2,310 @@
-* [DataCamp](https://www.datacamp.com/), [Codecademy](https://www.codecademy.com/) and [Coursera](https://www.coursera.org/) host a range of introductory courses, but more advanced courses are often not free.
+-- Guido van Rossum (creator of Python)
+```
 
-Please note that non-government analysis learning may not follow good practices. However, exposure to a range of applied examples will still benefit your learning.
-You should compare and contrast your learning to the good practices outlined in this section.
+When writing non-trivial code it is wise to assume that at some point someone else will need to understand, use and adapt your code. It might be yourself in six months time - after you've forgotten how the code works, or its design intent. Therefore every time you write such code, it is incredibly important to empathise with these potential users and produce code that is tidy, understandable and does not add unnecessary complexity.
+
+Common barriers to writing readable codebases include documentation that is hard to understand or absent, walls of code with repeating functionality that is hard to absorb in 'chunks' or overcomplicated solutions that solve the problem in ways that could be simplified. Avoiding these issues is essential to make sure that your analysis is reproducible, auditable and assured. Therefore it is our professional responsibility to avoid putting such barriers in place whenever possible.
+
+This chapter highlights some good coding practices that will improve the readability and maintainability of your code.
+Here, readability refers how easily another analyst can gain a decent understand of how your code works, within a reasonable amount of time.
+Maintainability refers to how easily other analysts can understand your code well enough to modify and repair it.
+
+(modular)=
+
+## Modular code
+
+Breaking your code down into smaller, more manageable chunks is a sensible way to improve readability. Regardless of the language, there are often methods to containerise your code into self-contained parts such as modules, classes or functions.
+
+### Functions
+
+When prototyping we often copy and paste code to 'make things work' but when time comes to wrap that work up, it is worth taking repetitivee code that can be easily parameterised and turning it into functions. Writing functions as well-sealed and reusable containers helps them be easily testable and readable.
+
+When starting to write functions consider what is the right **level of abstraction**. Namely, can this large piece of code be turned into concise and readable function as it is, without having to pass too many arguments to the resulting function? If not, perhaps you need to break the code into smaller helper functions (that can also be reused in other places across the codebase) and then **use these smaller functions to build up a larger function that performs the actions you need**.
```

Comment:
no bolding unless really necessary - it's inaccessible

Reply:
Looks like Markdown (or at least jupyter-book) uses the `<strong>` tag for bolding, which is semantic and screen reader friendly. We should still use it sparingly, but we should be fine to use it when it's necessary to make a point - for future reference

Reply:
good to know - i take it back!

Reply:
I had to look it up, cause I assumed the same 😄